csharp/src/Ice/InputStream-Class.cs

```diff
@@ -292,93 +321,45 @@ private AnyClass ReadInstance(int index)
-                }
-                else if (_current.SliceCompactId.HasValue)
-                {
-                    cls = Communicator.ResolveCompactId(_current.SliceCompactId.Value);
-                }
-
-                if (cls != null)
-                {
-                    try
+                    IClassFactory? factory = null;
+                    if (typeId != null)
                     {
-                        Debug.Assert(!cls.IsAbstract && !cls.IsInterface);
-                        v = (AnyClass?)Activator.CreateInstance(cls);
+                        Debug.Assert(_current.SliceCompactId == null);
+                        factory = Communicator.FindClassFactory(typeId);
                     }
-                    catch (Exception ex)
+                    else if (_current.SliceCompactId is int compactId)
                     {
-                        string typeIdString = typeId ?? _current.SliceCompactId!.ToString()!;
-                        throw new InvalidDataException(@$"failed to create an instance of type `{cls.Name
-                            } while reading a class with type ID {typeIdString}", ex);
+                        factory = Communicator.FindClassFactory(compactId);
                     }
-                }
-
-                if (v != null)
-                {
-                    // We have an instance, get out of this loop.
-                    break;
-                }
 
-                // Slice off what we don't understand, and save the indirection table (if any) in
-                // deferredIndirectionTableList.
-                deferredIndirectionTableList ??= new List<int>();
-                deferredIndirectionTableList.Add(SkipSlice());
-
-                // If this is the last slice, keep the instance as an opaque UnknownSlicedClass object.
-                if ((_current.SliceFlags & EncodingDefinitions.SliceFlags.IsLastSlice) != 0)
-                {
-                    v = new UnknownSlicedClass();
-                    break;
-                }
-
-                typeId = ReadSliceHeaderIntoCurrent(); // Read next Slice header for next iteration.
-            }
-
-            if (++_classGraphDepth > Communicator.ClassGraphDepthMax)
-            {
-                throw new InvalidDataException("maximum class graph depth reached");
-            }
+                    if (factory != null)
+                    {
+                        return factory.Read(this);
+                    }
 
-            // Add the instance to the map/list of instances. This must be done before reading the instances (for
-            // circular references).
-            _instanceMap ??= new List<AnyClass>();
-            _instanceMap.Add(v);
+                    // Slice off what we don't understand, and save the indirection table (if any) in
+                    // DeferredIndirectionTableList.
+                    _current.DeferredIndirectionTableList ??= new List<int>();
```

Comment:
I think I would put the murky details of slice skipping into the `SkipSlice` implementation and move the `DefferedIndirectTableList` setting into the `SkipSlice` implementation. 

`SkipSlice` could return a boolean indicating whether or not this is the last slice.

It's a detail but I also find the try/finally a bit misleading as it's not really used for exception handling... the loop could be re-arranged to not be a `while(true)` loop I guess.

Reply:
Refactored as suggested.